packages/components/src/FilterBar/FilterBar.spec.tsx

```diff
@@ -752,4 +754,132 @@ describe("<FilterBar />", () => {
+      department: string[]
+      manager: string[]
+    }
+
+    const MockFilterAsyncComponent = ({
+      id,
+      fetcher,
+    }: {
+      id: string
+      fetcher: (args: Partial<AsyncValues>) => Promise<Items>
+    }): JSX.Element => {
+      const [items, setItems] = useState<Items>([])
+      const { getActiveFilterValues } = useFilterBarContext()
+      const activeFilterVals = getActiveFilterValues()
+
+      useEffect(() => {
+        fetcher(activeFilterVals).then(fetchedItems => {
+          if (JSON.stringify(fetchedItems) !== JSON.stringify(items)) {
+            setItems(fetchedItems)
+          }
+        })
+      }, [JSON.stringify(activeFilterVals)])
+
+      return (
+        <FilterBar.MultiSelect id={id} items={items}>
+          {() => (
+            <FilterMultiSelect.ListBox>
+              {({ allItems }) => (
+                <FilterMultiSelect.ListBoxSection
+                  items={allItems}
+                  sectionName="All Items"
+                >
+                  {item => (
+                    <FilterMultiSelect.Option key={item.key} item={item} />
+                  )}
+                </FilterMultiSelect.ListBoxSection>
+              )}
+            </FilterMultiSelect.ListBox>
+          )}
+        </FilterBar.MultiSelect>
+      )
+    }
+
+    const fetchDepartmentOptions = jest.fn(() =>
+      Promise.resolve([
+        { value: "1", label: "Engineering" },
+        { value: "2", label: "Design" },
+      ])
+    )
+
+    const fetchManagerOptions = jest.fn(() =>
+      Promise.resolve([
+        { value: "3", label: "Superman" },
+        { value: "4", label: "Batman" },
+      ])
+    )
+
+    const config = [
+      {
+        id: "department",
+        name: "Department",
+        Component: (
+          <MockFilterAsyncComponent
+            id="department"
+            fetcher={fetchDepartmentOptions}
+          />
+        ),
+      },
+      {
+        id: "manager",
+        name: "Manager",
+        Component: (
+          <MockFilterAsyncComponent
+            id="manager"
+            fetcher={fetchManagerOptions}
+          />
+        ),
+      },
+    ] satisfies Filters<AsyncValues>
+
+    it("An async component can re-fetch with all active filter values pulled off of the FilterBarContext", async () => {
+      const { getByText, queryByText } = render(
+        <FilterBarWrapper<AsyncValues> filters={config} defaultValues={{}} />
+      )
+
+      // open department filter
+      user.click(getByText("Department"))
+
+      // select engineering
+      await waitFor(() => {
+        expect(queryByText("Engineering")).toBeInTheDocument()
+        expect(queryByText("Design")).toBeInTheDocument()
+      })
+
+      user.click(getByText("Engineering"))
+
+      // assert last fetch manager call has selected department value in it.
+      await waitFor(() => {
+        expect(fetchManagerOptions.mock.lastCall).toEqual([
+          { department: ["1"] },
+        ])
+      })
```

Comment:
Given this is moreso a behavioural test, instead of asserting against functions being called, awaiting a change in the UI is a better check.

I was a bit confused reading through at the start, jumping between seeing steps based around Department, then reading the expect being based around Manager, but having it equal to something from Department.

The actual technical implementation does not matter as much as achieving the outcome, which is what this test should be based around.

Reply:
Cool. I've updated the implementation of the jest fetcher to have a bit of logic in there to return different results based on what filter values are present. Pretty simple but it does the job. This then lets me do assertions on what's actually in the UI, rather than observe the jest spies.